File: e-commerceiosApp/e-commerceiosApp/ViewModels/AdminCJImportViewModel.swift
import Foundation
import SwiftUI

@MainActor
class AdminCJImportViewModel: ObservableObject {
    // Search state
    @Published var searchQuery = ""
    @Published var selectedCategory = "All"
    @Published var selectedWarehouse: String? = nil
    @Published var selectedShippingCountry: String? = nil
    @Published var sortByTrending = false
    @Published var isInitialLoad = true
    @Published var searchResults: [CJProduct] = []
    @Published var isLoading = false
    @Published var isLoadingMore = false
    @Published var totalResults = 0
    @Published var currentPage = 1
    @Published var errorMessage: String?
    
    // Import state
    @Published var selectedProduct: CJProduct?
    @Published var isImporting = false
    @Published var importingProductId: String?
    @Published var importedProductIds: Set<String> = []
    @Published var showImportDialog = false
    @Published var showImportSuccess = false
    @Published var lastImportedProduct: CJImportResponse?
    @Published var importError: String?
    
    // Import form fields
    @Published var sellingPrice = ""
    @Published var commissionRate = "10"
    @Published var customDescription = ""
    
    let categories = ["All", "Electronics", "Clothing", "Home & Garden", "Beauty", "Sports", "Toys", "Jewelry", "Automotive", "Pet Supplies"]
    let warehouses: [(label: String, value: String?)] = [
        ("All", nil), ("🇨🇳 China", "CN"), ("🇺🇸 USA", "US"), ("🇩🇪 Europe", "EU")
    ]
    let shippingCountries: [(label: String, value: String?)] = [
        ("All", nil), ("🇺🇸 US", "US"), ("🇫🇷 FR", "FR"), ("🇲🇦 MA", "MA"),
        ("🇩🇿 DZ", "DZ"), ("🇨🇦 CA", "CA"), ("🇬🇧 GB", "GB"), ("🇩🇪 DE", "DE"), ("🇧🇪 BE", "BE")
    ]
    
    var hasMorePages: Bool { searchResults.count < totalResults }
    
    // MARK: - Search
    
    func searchProducts() async {
        guard !searchQuery.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        isLoading = true
        errorMessage = nil
        currentPage = 1
        searchResults = []
        do {
            let response = try await AdminApiService.shared.searchCJProducts(
                query: searchQuery,
                category: selectedCategory == "All" ? nil : selectedCategory,
                page: 1,
                warehouse: selectedWarehouse,
                shippingCountry: selectedShippingCountry,
                sortBy: sortByTrending ? "trending" : nil
            )
            searchResults = response.products
            totalResults = response.total
            currentPage = response.page
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }
    
    func loadTrendingProducts() async {
        guard isInitialLoad else { return }
        isLoading = true
        do {
            let response = try await AdminApiService.shared.searchCJProducts(
                query: "trending",
                sortBy: "trending"
            )
            searchResults = response.products
            totalResults = response.total
            currentPage = response.page
        } catch {
            // Silently fail — welcomeView shows as fallback
        }
        isLoading = false
        isInitialLoad = false
    }
    
    func loadNextPage() async {
        guard hasMorePages, !isLoadingMore else { return }
        isLoadingMore = true
        let nextPage = currentPage + 1
        do {
            let response = try await AdminApiService.shared.searchCJProducts(
                query: searchQuery.isEmpty ? "trending" : searchQuery,
                category: selectedCategory == "All" ? nil : selectedCategory,
                page: nextPage,
                warehouse: selectedWarehouse,
                shippingCountry: selectedShippingCountry,
                sortBy: (searchQuery.isEmpty || sortByTrending) ? "trending" : nil
            )
            searchResults.append(contentsOf: response.products)
            totalResults = response.total
            currentPage = response.page
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoadingMore = false
    }
    
    // MARK: - Filter setters
    
    func selectCategory(_ category: String) {
        selectedCategory = category
        if !searchQuery.isEmpty {
            Task { await searchProducts() }
        }
    }
    
    func setWarehouse(_ warehouse: String?) {
        selectedWarehouse = warehouse
        if !searchQuery.isEmpty { Task { await searchProducts() } }
    }
    
    func setShippingCountry(_ country: String?) {
        selectedShippingCountry = country
        if !searchQuery.isEmpty { Task { await searchProducts() } }
    }
    
    func toggleTrending() {
        sortByTrending.toggle()
        if !searchQuery.isEmpty { Task { await searchProducts() } }
    }
    
    // MARK: - Import
    
    func selectProduct(_ product: CJProduct) {
        selectedProduct = product
        sellingPrice = String(format: "%.2f", product.sellPrice * 1.5)
        commissionRate = "10"
        customDescription = ""
        importError = nil
        showImportDialog = true
    }
    
    func importProduct() async {
        guard let product = selectedProduct else { return }
        guard let price = Double(sellingPrice), price > 0 else {
            importError = "Please enter a valid selling price"
            return
        }
        guard let commission = Double(commissionRate), commission >= 0, commission <= 100 else {
            importError = "Commission rate must be between 0 and 100"
            return
        }
        isImporting = true
        importingProductId = product.productId
        importError = nil
        do {
            let request = CJImportRequest(
                cjProductId: product.productId,
                cjVariantId: product.variants?.first?.variantId,
                commissionRate: commission,
                categoryId: nil,
                customDescription: customDescription.isEmpty ? nil : customDescription,
                sellingPrice: price
            )
            let response = try await AdminApiService.shared.importCJProduct(request: request)
            lastImportedProduct = response
            importedProductIds.insert(product.productId)
            showImportDialog = false
            showImportSuccess = true
        } catch {
            importError = error.localizedDescription
        }
        isImporting = false
        importingProductId = nil
    }
    
    func dismissImportSuccess() {
        showImportSuccess = false
        lastImportedProduct = nil
    }
    
    func clearSearch() {
        searchQuery = ""
        searchResults = []
        totalResults = 0
        currentPage = 1
    }
}

File: e-commerceAndroidApp/src/main/java/com/project/e_commerce/android/presentation/viewModel/admin/AdminCJImportViewModel.kt
package com.project.e_commerce.android.presentation.viewModel.admin

import android.content.Context
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.project.e_commerce.android.data.api.CJImportResponse
import com.project.e_commerce.android.data.api.CJProduct
import com.project.e_commerce.android.data.repository.AdminRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel pour l'import de produits CJ Dropshipping
 * 
 * Permet aux admins de :
 * - Rechercher des produits sur CJ Dropshipping
 * - Importer des produits dans le catalogue
 * - Synchroniser les prix/stocks avec CJ
 */
class AdminCJImportViewModel(
    private val repository: AdminRepository,
    private val context: Context
) : ViewModel() {

    private val _uiState = MutableStateFlow(AdminCJImportUiState())
    val uiState: StateFlow<AdminCJImportUiState> = _uiState.asStateFlow()

    private val encryptedPrefs by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        EncryptedSharedPreferences.create(
            context,
            "admin_secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    private fun getAdminToken(): String? {
        return encryptedPrefs.getString("admin_token", null)
    }

    /**
     * Recherche des produits sur CJ Dropshipping
     */
    fun searchProducts(
        query: String,
        category: String? = null,
        warehouse: String? = _uiState.value.selectedWarehouse,
        shippingCountry: String? = _uiState.value.selectedShippingCountry,
        sortBy: String? = if (_uiState.value.sortByTrending) "trending" else null
    ) {
        if (query.isBlank()) {
            _uiState.value = _uiState.value.copy(error = "Please enter a search query")
            return
        }

        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(
                    isLoading = true,
                    error = null,
                    searchQuery = query,
                    selectedCategory = category
                )
                
                val token = getAdminToken() ?: throw Exception("Not logged in as admin")
                
                val response = repository.searchCJProducts(
                    token = token,
                    query = query,
                    category = category,
                    page = 1,
                    warehouse = warehouse,
                    shippingCountry = shippingCountry,
                    sortBy = sortBy
                )
                
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    searchResults = response.products,
                    totalResults = response.total,
                    currentPage = response.page,
                    error = null
                )
                
                Log.d("AdminCJImportVM", "✅ Found ${response.total} CJ products for '$query'")
            } catch (e: Exception) {
                Log.e("AdminCJImportVM", "❌ Error searching CJ products", e)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to search CJ products"
                )
            }
        }
    }

    /**
     * Charge les produits tendance au lancement (état initial)
     */
    fun loadTrendingProducts() {
        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(isLoading = true, error = null)
                val token = getAdminToken() ?: return@launch
                val response = repository.searchCJProducts(
                    token = token,
                    query = "trending",
                    sortBy = "trending"
                )
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    searchQuery = "",
                    searchResults = response.products,
                    totalResults = response.total,
                    currentPage = response.page,
                    isInitialLoad = false,
                    error = null
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, isInitialLoad = false)
            }
        }
    }

    /** Applique le filtre entrepôt et relance la recherche si active */
    fun setWarehouse(warehouse: String?) {
        _uiState.value = _uiState.value.copy(selectedWarehouse = warehouse)
        if (_uiState.value.searchQuery.isNotBlank()) {
            searchProducts(_uiState.value.searchQuery, _uiState.value.selectedCategory, warehouse)
        }
    }

    /** Applique le filtre pays expédition et relance la recherche si active */
    fun setShippingCountry(country: String?) {
        _uiState.value = _uiState.value.copy(selectedShippingCountry = country)
        if (_uiState.value.searchQuery.isNotBlank()) {
            searchProducts(_uiState.value.searchQuery, _uiState.value.selectedCategory,
                shippingCountry = country)
        }
    }

    /** Bascule le tri tendance et relance la recherche si active */
    fun toggleTrending() {
        val newVal = !_uiState.value.sortByTrending
        _uiState.value = _uiState.value.copy(sortByTrending = newVal)
        if (_uiState.value.searchQuery.isNotBlank()) {
            searchProducts(_uiState.value.searchQuery, _uiState.value.selectedCategory,
                sortBy = if (newVal) "trending" else null)
        }
    }

    /**
     * Charge la page suivante des résultats
     */
    fun loadNextPage() {
        val currentState = _uiState.value
        if (currentState.isLoading || currentState.searchQuery.isBlank()) return
        
        // Check if there are more pages
        val totalPages = (currentState.totalResults + 19) / 20 // page_size = 20
        if (currentState.currentPage >= totalPages) return

        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(isLoadingMore = true)
                
                val token = getAdminToken() ?: throw Exception("Not logged in as admin")
                
                val response = repository.searchCJProducts(
                    token = token,
                    query = currentState.searchQuery,
                    category = currentState.selectedCategory,
                    page = currentState.currentPage + 1,
                    warehouse = currentState.selectedWarehouse,
                    shippingCountry = currentState.selectedShippingCountry,
                    sortBy = if (currentState.sortByTrending) "trending" else null
                )
                
                _uiState.value = _uiState.value.copy(
                    isLoadingMore = false,
                    searchResults = currentState.searchResults + response.products,
                    currentPage = response.page,
                    error = null
                )
                
                Log.d("AdminCJImportVM", "✅ Loaded page ${response.page}")
            } catch (e: Exception) {
                Log.e("AdminCJImportVM", "❌ Error loading next page", e)
                _uiState.value = _uiState.value.copy(
                    isLoadingMore = false,
                    error = e.message ?: "Failed to load more products"
                )
            }
        }
    }

    /**
     * Importe un produit CJ dans le catalogue
     */
    fun importProduct(
        cjProductId: String,
        cjVariantId: String? = null,
        commissionRate: Double = 10.0,
        categoryId: String? = null,
        customDescription: String? = null,
        sellingPrice: Double? = null
    ) {
        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(
                    isImporting = true,
                    importingProductId = cjProductId,
                    importError = null
                )
                
                val token = getAdminToken() ?: throw Exception("Not logged in as admin")
                
                val response = repository.importCJProduct(
                    token = token,
                    cjProductId = cjProductId,
                    cjVariantId = cjVariantId,
                    commissionRate = commissionRate,
                    categoryId = categoryId,
                    customDescription = customDescription,
                    sellingPrice = sellingPrice
                )
                
                // Add to imported products list
                val importedProducts = _uiState.value.importedProducts + response
                
                _uiState.value = _uiState.value.copy(
                    isImporting = false,
                    importingProductId = null,
                    importedProducts = importedProducts,
                    showImportSuccess = true,
                    lastImportedProduct = response,
                    importError = null
                )
                
                Log.d("AdminCJImportVM", "✅ Imported CJ product: ${response.name}")
            } catch (e: Exception) {
                Log.e("AdminCJImportVM", "❌ Error importing CJ product", e)
                _uiState.value = _uiState.value.copy(
                    isImporting = false,
                    importingProductId = null,
                    importError = e.message ?: "Failed to import product"
                )
            }
        }
    }

    /**
     * Synchronise un produit avec CJ (met à jour prix/stock)
     */
    fun syncProduct(productId: String) {
        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(
                    isSyncing = true,
                    syncingProductId = productId
                )
                
                val token = getAdminToken() ?: throw Exception("Not logged in as admin")
                
                val response = repository.syncCJProduct(token, productId)
                
                _uiState.value = _uiState.value.copy(
                    isSyncing = false,
                    syncingProductId = null,
                    syncMessage = response.message
                )
                
                Log.d("AdminCJImportVM", "✅ Synced product: ${response.message}")
            } catch (e: Exception) {
                Log.e("AdminCJImportVM", "❌ Error syncing product", e)
                _uiState.value = _uiState.value.copy(
                    isSyncing = false,
                    syncingProductId = null,
                    error = e.message ?: "Failed to sync product"
                )
            }
        }
    }

    /**
     * Sélectionne un produit pour voir les détails / configurer l'import
     */
    fun selectProduct(product: CJProduct?) {
        _uiState.value = _uiState.value.copy(selectedProduct = product)
    }

    /**
     * Ferme le dialogue de succès d'import
     */
    fun dismissImportSuccess() {
        _uiState.value = _uiState.value.copy(
            showImportSuccess = false,
            lastImportedProduct = null
        )
    }

    /**
     * Efface l'erreur d'import
     */
    fun clearImportError() {
        _uiState.value = _uiState.value.copy(importError = null)
    }

    /**
     * Efface l'erreur générale
     */
    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    /**
     * Réinitialise la recherche
     */
    fun clearSearch() {
        _uiState.value = AdminCJImportUiState()
    }
}

/**
 * UI State pour AdminCJImportScreen
 */
data class AdminCJImportUiState(
    // Search state
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val searchQuery: String = "",
    val selectedCategory: String? = null,
    val selectedWarehouse: String? = null,
    val selectedShippingCountry: String? = null,
    val sortByTrending: Boolean = false,
    val isInitialLoad: Boolean = true,
    val searchResults: List<CJProduct> = emptyList(),
    val totalResults: Int = 0,
    val currentPage: Int = 1,
    
    // Product selection
    val selectedProduct: CJProduct? = null,
    
    // Import state
    val isImporting: Boolean = false,
    val importingProductId: String? = null,
    val importedProducts: List<CJImportResponse> = emptyList(),
    val showImportSuccess: Boolean = false,
    val lastImportedProduct: CJImportResponse? = null,
    val importError: String? = null,
    
    // Sync state
    val isSyncing: Boolean = false,
    val syncingProductId: String? = null,
    val syncMessage: String? = null,
    
    // General error
    val error: String? = null
)

/**
 * Categories CJ pour filtrage
 */
object CJCategories {
    val categories = listOf(
        "All" to null,
        "Electronics" to "electronics",
        "Clothing" to "clothing",
        "Home & Garden" to "home_garden",
        "Beauty" to "beauty",
        "Sports" to "sports",
        "Toys" to "toys",
        "Jewelry" to "jewelry",
        "Automotive" to "automotive",
        "Pet Supplies" to "pet_supplies"
    )
}

/**
 * Entrepôts CJ disponibles
 */
object CJWarehouses {
    val warehouses = listOf(
        "All" to null,
        "🇨🇳 China (CN)" to "CN",
        "🇺🇸 USA (US)" to "US",
        "🇩🇪 Europe (EU)" to "EU"
    )
}

/**
 * Pays de livraison supportés
 */
object CJShippingCountries {
    val countries = listOf(
        "All" to null,
        "🇺🇸 United States" to "US",
        "🇫🇷 France" to "FR",
        "🇲🇦 Morocco" to "MA",
        "🇩🇿 Algeria" to "DZ",
        "🇨🇦 Canada" to "CA",
        "🇬🇧 United Kingdom" to "GB",
        "🇩🇪 Germany" to "DE",
        "🇧🇪 Belgium" to "BE"
    )
}
